### PR TITLE
Feat kind support map pinning v3

### PR DESCRIPTION
**Pull request by maryamtahhan**
Rebasing the previous [PR](https://github.com/intel/afxdp-plugins-for-kubernetes/pull/59) to support bpf map pinning after Kind support was merged to main.

Going to close [PR 59](https://github.com/intel/afxdp-plugins-for-kubernetes/pull/59)

Will transition from draft after some local testing.

### CODE REVIEW

**Comment by maryamtahhan** [comment]
Tested with a kind cluster.

configure the unprivileged bpf kernel flag on the kind worker nodes

```bash
$ docker exec af-xdp-deployment-worker sysctl kernel.unprivileged_bpf_disabled=0
kernel.unprivileged_bpf_disabled = 0
$ docker exec af-xdp-deployment-worker2 sysctl kernel.unprivileged_bpf_disabled=0
```
 
 Used the following NAD:

 ```yaml
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: afxdp-network
  annotations:
    k8s.v1.cni.cncf.io/resourceName: afxdp/myPool
spec:
  config: '{
      "cniVersion": "0.3.0",
      "type": "afxdp",
      "mode": "primary",
      "logFile": "afxdp-cni.log",
      "logLevel": "debug",
      "dpSyncer": true,
      "ipam": {
        "type": "host-local",
        "subnet": "192.168.1.0/24",
        "rangeStart": "192.168.1.200",
        "rangeEnd": "192.168.1.220",
        "routes": [
          { "dst": "0.0.0.0/0" }
        ],
        "gateway": "192.168.1.1"
      }
    }'
 ```
 
 and the following pod spec:
 ```
 apiVersion: v1
kind: Pod
metadata:
  name: cndp-0-0
  annotations:
    k8s.v1.cni.cncf.io/networks: afxdp-network 
spec:
  containers:
    - name: cndp-0
      command: ["/bin/bash"]
      args: ["-c", "./jsonc_gen.sh -kp ; cndpfwd -c config.jsonc lb;"] # to run in a kind cluster using bpf map pinning
      image:  quay.io/mtahhan/cndp-map-pinning:latest 
      imagePullPolicy: IfNotPresent
      resources:
        requests:
          afxdp/myPool: '1'
        limits:
          afxdp/myPool: '1
 ```
 
 Then creating and deleting the cndp pod - the logs of the Device plugin are updated accordingly with bpf map pinning messages/information

**Comment by maryamtahhan** [comment]
I just spotted the 

```
libbpf: can't get next link: Operation not permitted
``` 

this is not expected... but doesn't block this PR at least... Let me see if CAP_BPF is the issue here... it should be either CAP_BPF or `unprivileged_bpf_disabled`...

**Comment by maryamtahhan** [comment]
Ok - I think we can ignore that warning... 
https://github.com/libbpf/libbpf/commit/8628610c322a it looks like it's a probe under the hood of libbpf and when bpf link support isn't detected it reverts back to netlink-based XDP prog... we indeed don't have permissions to make this bpf call is the pod is unprivileged... 
I did notice however that even CAP_BPF didn't provide enough privilege for this call...

**Comment by garyloug** [comment]
Hey @maryamtahhan, we've seen `libbpf: can't get next link: Operation not permitted` before. It's pod privileges we think, running as root fixed it, but obviously that's not the solution.

ok so it's a probe for libbpf rather than breaking functionality? Or is it breaking functionality?

Is it something we could ask the DP to configure for us?

Kind of related:
Client being created in #65 and we'll put a C wrapper on this once finalised.

**Comment by johnoloughlin** [comment]
capabilities don't get added to a non root users shell. You need to use setcap on the specific binary that needs the capability in the dockerfile (you cant do it in a running container). Then you also need to have the matching capability in the pod spec.
you can easily check the capabilities of the current shell with capsh --print.
getcap can be used to check  specific capability of a binary

**Comment by maryamtahhan** [comment]
> capabilities don't get added to a non root users shell. You need to use setcap on the specific binary that needs the capability in the dockerfile (you cant do it in a running container). Then you also need to have the matching capability in the pod spec. you can easily check the capabilities of the current shell with capsh --print. getcap can be used to check specific capability of a binary

sorry, what's the context here?

**Comment by johnoloughlin** [comment]
My mistake, I thought that you were running as a non root user from grays comment:
"libbpf: can't get next link: Operation not permitted before. It's pod privileges we think, running as root fixed it" 
such as https://github.com/intel/afxdp-plugins-for-kubernetes/blob/main/test/e2e/pod-1c1d.yaml#L9 where you `runAsUser: 1500` or that you had baked the non root user into the dockerfile.
I realise now that he meant unprivileged root user

File: examples/nad_with_syncer.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  annotations:
+    k8s.v1.cni.cncf.io/resourceName: afxdp/myPool            # Needs to match the device plugin pool name / resource type
+spec:
+  config: '{
+      "cniVersion": "0.3.0",
+      "type": "afxdp",                                       # CNI binary, leave as afxdp
+      "mode": "primary",                                     # CNI mode setting (required)
+      "logFile": "afxdp-cni.log",                            # CNI log file location (optional)
+      "logLevel": "debug",                                   # CNI logging level (optional)
+      "dpSyncer": true,                                      # Sync with Device Plugin over gRPC (optional)
```
Comment (maryamtahhan) [inline examples/nad_with_syncer.yaml:16]: should make this more explicit that it's only for bpfmap pinning or on the DP side if pinning isn't enabled then just return success...

File: internal/dpcnisyncerserver/server.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+				return
+			}
+		}
+	}
+
+	s.mapManagers = append(s.mapManagers, b)
+}
+
+func (s *SyncerServer) DelNetDev(ctx context.Context, in *pb.DeleteNetDevReq) (*pb.DeleteNetDevResp, error) {
+	netDevName := in.GetName()
```
Comment (maryamtahhan) [inline internal/dpcnisyncerserver/server.go:64]: should check if bpfmappinning is configured in case of misconfiguration on the CNI side

**Review by garyloug — DISMISSED** [review]

**Comment by maryamtahhan** [comment]
I've rebased on main and tested in Kind... everything is working as expected.

**Review by garyloug — APPROVED** [review]

**Pull request merged**